In [1]:
import pandas as pd
import numpy as np
from scipy.stats import norm
from scipy.optimize import brentq, least_squares
import matplotlib.pylab as plt
import math


In [2]:
def impliedVolatility(S, K, r, price, T, payoff):
    try:
        if (payoff.lower() == 'call'):
            impliedVol = brentq(lambda x: price -
                                BlackScholesLognormalCall(S, K, r, x, T),
                                1e-12, 10.0)
        elif (payoff.lower() == 'put'):
            impliedVol = brentq(lambda x: price -
                                BlackScholesLognormalPut(S, K, r, x, T),
                                1e-12, 10.0)
        else:
            raise NameError('Payoff type not recognized')
    except Exception:
        impliedVol = np.nan

    return impliedVol

def preprocess_options_data(file_path):
    """
    Preprocess options data by calculating mid-price, scaling strike prices, 
    and mapping call/put flags.

    Parameters:
    - file_path (str): Path to the CSV file containing options data.

    Returns:
    - DataFrame: Processed options data with 'mid', 'strike', and 'payoff' columns.
    """
    df = pd.read_csv(file_path)
    df['mid'] = 0.5 * (df['best_bid'] + df['best_offer'])
    df['strike'] = df['strike_price'] * 0.001  # Adjusting strike price scale
    df['payoff'] = df['cp_flag'].map(lambda x: 'call' if x == 'C' else 'put')
    return df

def BlackScholesLognormalCall(S, K, r, sigma, T):
    d1 = (np.log(S/K)+(r+sigma**2/2)*T) / (sigma*np.sqrt(T))
    d2 = d1 - sigma*np.sqrt(T)
    return S*norm.cdf(d1) - K*np.exp(-r*T)*norm.cdf(d2)

def BlackScholesLognormalPut(S, K, r, sigma, T):
    d1 = (np.log(S/K)+(r+sigma**2/2)*T) / (sigma*np.sqrt(T))
    d2 = d1 - sigma*np.sqrt(T)
    return K*np.exp(-r*T)*norm.cdf(-d2) - S*norm.cdf(-d1)

In [15]:

df_rates = pd.read_csv("zero_rates_20201201.csv")\
            .sort_values('days')
df_spx = preprocess_options_data("SPX_options.csv")
df_spy = preprocess_options_data("SPY_options.csv")

S_spx = 3662.45
S_spy = 366.02
exdate = 20210115
days_to_expiry = (pd.Timestamp(str(exdate)) - pd.Timestamp('2020-12-01')).days
r = np.interp(days_to_expiry, df_rates['days'].values, df_rates['rate'].values)
T = days_to_expiry / 365


# QF620 Project Part 3
Suppose on 1-Dec-2020, we need to evaluate an exotic European derivative
expiring on 15-Jan-2021
## 1. Contract Payoff function
Payoff function is

$$
S^{\frac{1}{3}}_T + 1.5 \times \log(S_T) + 10.0

$$


### 1.1 Black-Scholes Model
under Black-Scholes

$$
S_T = S_0 e^{(r - \frac{\sigma^2}{2})T + \sigma W_T}

$$

Therefore

$$
\begin{aligned}

V_0 
&= e^{-rT} \mathbb{E}
\left[
    {S_0 e^{\frac{1}{3} (r - \frac{\sigma^2}{2})T + \frac{1}{3} \sigma W_T}}
    + 1.5 \times \log(S_0 e^{(r - \frac{\sigma^2}{2})T 
    + \sigma W_T}) + 10.0
\right] \\
&= 
e^{-rT}
\left[
{S_0 e^{\frac{1}{3} (r - \frac{\sigma^2}{2})T}} \mathbb{E}[{e^{\frac{1}{3} \sigma W_T}}] +

	 1.5 \left[ 
		 \log{S_0} + \left(r - \frac{\sigma^2}{2} \right) T  + \mathbb{E}[\sigma W_T] 
	\right]
	+
	10.0
\right]
\\\\
V_0
&= 
e^{-rT}
\left[
{S_0 e^{\frac{1}{3} (r - \frac{\sigma^2}{3})T}} +

	 1.5 \left[ 
		 \log{S_0} + \left(r - \frac{\sigma^2}{2} \right) T 
	\right]
	+
	10.0
\right]
\\

\end{aligned}
$$


#### 1.1.1 $\sigma$ used
There is many market implied volatilities 
We can try using ATM strike volatility to price as a risk neutral measure where $K = S_0 e^{rT}$

##### 1.1.1.1 SPX 

In [16]:
F = S_spx * np.exp(r * T)
# Filter for the current expiration date
df_ex = df_spx[df_spx["exdate"] == exdate].copy()

In [17]:
df_spx[df_spx['exdate']==20210115]

date    exdate cp_flag  strike_price  best_bid  best_offer  \
822   20201201  20210115       C        100000    3544.8      3567.7   
823   20201201  20210115       C        200000    3444.9      3467.8   
824   20201201  20210115       C        300000    3345.0      3367.9   
825   20201201  20210115       C        400000    3245.0      3267.9   
826   20201201  20210115       C        500000    3145.1      3168.0   
...        ...       ...     ...           ...       ...         ...   
1555  20201201  20210115       P       5000000    1330.8      1346.6   
1556  20201201  20210115       P       5100000    1428.3      1451.2   
1557  20201201  20210115       P       5200000    1528.3      1551.2   
1558  20201201  20210115       P       5300000    1628.2      1651.1   
1559  20201201  20210115       P       5400000    1728.1      1751.0   

     exercise_style      mid  strike payoff  
822               E  3556.25   100.0   call  
823               E  3456.35   200.0   call  
824               E  3356.45   300.0   call  
825               E  3256.45   400.0   call  
826               E  3156.55   500.0   call  
...             ...      ...     ...    ...  
1555              E  1338.70  5000.0    put  
1556              E  1439.75  5100.0    put  
1557              E  1539.75  5200.0    put  
1558              E  1639.65  5300.0    put  
1559              E  1739.55  5400.0    put  

[738 rows x 10 columns]

In [36]:
# Filter and collect strikes and implied vols
df_ex['vols'] = df_ex.apply(lambda x: impliedVolatility(S_spx, x['strike'], r, x['mid'], T, x['payoff']), axis=1)
df_ex.dropna(inplace=True)
call_df, put_df = df_ex[df_ex['payoff'] == 'call'], df_ex[df_ex['payoff'] == 'put']
strikes = put_df['strike'].values
impliedvols = [
    call_df[call_df['strike'] == K]['vols'].values[0] 
        if K > S_spx else put_df[put_df['strike'] == K]['vols'].values[0]
            for K in strikes
]
df_iv = pd.DataFrame({'strike': strikes, 'impliedvol': impliedvols})
closest_strike = df_iv['strike'].iloc[(df_iv['strike'] - F).abs().idxmin()]

# get pricing vol for contract
spx_ATM_implied_vol = df_iv[df_iv['strike'] == closest_strike]['impliedvol'].values[0]

In [53]:
np.exp(-r * T) *\
    (
        S_spx * np.exp(1/3 * (r - (spx_ATM_implied_vol**2) / 3) * T) \
        + 1.5 * (np.log(S_spx) + (r - (spx_ATM_implied_vol**2) / 2) * T)
        + 10
	)


3622.5922491209762

In [34]:
spx_ATM_implied_vol

0.09239431532592252

In [35]:
df_iv.describe()

strike  impliedvol
count   369.000000  369.000000
mean   3025.149051    0.403861
std     854.310548    0.312978
min     100.000000    0.053265
25%    2575.000000    0.273486
50%    3145.000000    0.346064
75%    3605.000000    0.482311
max    5400.000000    3.043075


#### 1.1.2 Valuation

### 1.2 Bachelier Model
#### 1.2.1 $\sigma$ used
#### 1.2.2 Valuation

### 1.2 Static Replication of European Payoff Model
*using SABR model calibrated from Part 2*
#### 1.2.1 Calibrated SABR model
SPX: Exdate 20210115: alpha = 1.817, rho = -0.404, nu = 2.790

SPY: Exdate 20210115: alpha = 0.908, rho = -0.489, nu = 2.729
#### 1.2.2 Valuation

## 2. Variance Swap
### 2.1 Black-Scholes Model
#### 2.1.1 $\sigma$ used
#### 2.1.2 Valuation

### 2.2 Bachelier Model
#### 2.2.1 $\sigma$ used
#### 2.2.2 Valuation

### 2.2 Static Replication of European Payoff Model
*using SABR model calibrated from Part 2*
#### 2.2.1 Calibrated SABR model
#### 2.2.2 Valuation